test/issue-2700/set-difference-spec.k

```diff
@@ -0,0 +1,31 @@
+module VERIFICATION
+    imports SET-DIFFERENCE
+endmodule
+
+module SET-DIFFERENCE-SPEC
+    imports VERIFICATION
+
+    claim [claim.diff.empty]:
+          <k> run(.Set -Set SetItem(_X)) => done(.Set) ... </k>
+
+    claim [claim.concat]:
+          <k> run((SetItem(X) REST) -Set SetItem(Y)) =>
+              done(SetItem(X) (REST -Set SetItem(Y))) ... </k>
+      requires notBool pred(X) andBool pred(Y)
+
+    claim [claim.diff.both.concrete.singletons]:
+          <k> run(SetItem(1) -Set SetItem(10)) => done(SetItem(1)) ... </k>
+
+    claim [claim.diff.left.concrete.singleton]:
+          <k> run(SetItem(1) -Set SetItem(X)) => done(SetItem(1)) ... </k>
+      requires pred(X)
+
+    // This claim does not pass, am I missing something?
+    // claim [claim.diff.left.symbolic.singleton]:
+    //       <k> run(SetItem(X) -Set SetItem(10)) => done(SetItem(X)) ... </k>
+    //   requires notBool pred(X)
```

Comment:
For the `concat` equation to apply, it requires `pred(10)` to be `true`. The `pred.false` equation doesn't apply for `pred(10)`, so it remains unevaluated.

I think adding `pred(10)` in the claim's requires should work, or adding another equation for `pred` which handles the other cases (or specifically the case for `10`).

Reply:
The LHS of this claim is not matched even if I remove `pred` altogether. It appears that this needs more investigation, I'll add some traces again tomorrow and see what I could find.

Reply:
@ana-pantilie I have added an even more reduced example to demonstrate the issue. Apparently a symbolic variable is not matched with a concrete value when it should. Here are the terms that I think should be matched, produced with a `trace` in `attemptEquation`, like we did before:
```
term1
/* Fl Fn D Spa */
LblSet'Coln'difference{}(
    /* Fl Fn D Spa */
    /* InternalSet: */ /* element: */ LblSetItem{}(
        /* Fl Fn D Spa */
        /* Inj: */ inj{SortInt{}, SortKItem{}}(
            /* Fl Fn D Sfa */ EquationVarX:SortInt{}
        )
    ),
    /* Fl Fn D Spa */
    /* InternalSet: */ /* element: */ LblSetItem{}(
        /* Fl Fn D Spa */
        /* Inj: */ inj{SortInt{}, SortKItem{}}(
            /* Fl Fn D Sfa */ EquationVarY:SortInt{}
        )
    )
)
term2
/* Fl Fn D Spa */
LblSet'Coln'difference{}(
    /* Fl Fn D Sfa */
    /* InternalSet: */ /* element: */ LblSetItem{}(
        /* Fl Fn D Sfa */
        /* Inj: */ inj{SortInt{}, SortKItem{}}(
            /* Fl Fn D Sfa */ ConfigVarX:SortInt{}
        )
    ),
    /* Fl Fn D Sfa Cl */
    /* InternalSet: */ /* concrete element: */ LblSetItem{}(
        /* Inj: */ inj{SortInt{}, SortKItem{}}(\dv{SortInt{}}("10"))
    )
)
```

Reply:
I don't think it should be handled by that case, could you add a unit test for this similar to the ones I added in the new matching case PR?

Reply:
Sure, will do.

By the way, if I change `10` to `Y`, the claim passes, i.e. the "structurally equal" works for variables.

Reply:
I added a [unit tests](https://github.com/kframework/kore/pull/2737/commits/2f5970aa138f33f7015b74135ce81517d105cd1d) that I think should correspond to the issue that arises in the [example](https://github.com/kframework/kore/pull/2737/commits/89265c1977cd172cee0bdf48a9ae8021e885cc9b).